enhancement(vrl): RFC for return expression

mladedav opened the pull request:
Partly based on #7496 but closer to a subset from #19766 which does not include any breaking changes such as adding a new reserved keyword to VRL.

<!--
**Your PR title must conform to the conventional commit spec!**
  <type>(<scope>)!: <description>

  * `type` = chore, enhancement, feat, fix, docs
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs", available scopes https://github.com/vectordotdev/vector/blob/master/.github/semantic.yml#L20
  * `description` = short description of the change

Examples:

  * enhancement(file source): Add `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fix a bug discovering new files
  * chore(external docs): Clarify `batch_size` option
-->

pront reviewed (commented):
Hi @mladedav, this RFC looks good. I see `return` as a valuable feature to have. Left one question.

mladedav commented:
Thanks, @pront! Sorry, but I don't see any question?

pront commented on rfcs/2023-02-08-7496-vrl-return.md:
```diff
@@ -0,0 +1,72 @@
+### User Experience
+
+- A `return` expression causes the VRL program to terminate, keeping any modifications made to the event.
+- The keyword is used by itself with no parameters and `.` will be sent to the output.
```
Worth considering returning an `expression`, it can be `.` or a variable or something that resolves to a constant. What do you think?

mladedav replied:
Yes, I think that is the end l goal to go for. I've tried to keep the scope small, but it can also be part of the RFC, with the current scope being probably the first implementation and having positional arguments being enhancement of that.

I opted for omitting it since it felt to me that writing
```
. = expr
return
```
is only slightly worse than
```
return expr
```

Do you think it's better to have it in this RFC or having it split as a later enhancement?

mladedav replied:
Also I think even if expressions are allowed, they should be optional, allowing simple `return` instead of the more explicit `return .`

But now I see that could be a contentious decision and some people can prefer that the language wants them to be explicit, so I'd like to also know your opinion on this.

pront replied:
I would take inspiration from https://doc.rust-lang.org/std/keyword.return.html so I would like to return an expression. See [here](https://github.com/vectordotdev/vector/blob/v0.36.0/src/conditions/vrl.rs#L101-L106), we can use both the returned value and the target `.`.

jszwedko replied:
To clarify a bit more: would `return foo` use the value of `foo` as the event to emit? Or would it still use `.`? If the latter, what would the "returned value" be used for?

mladedav replied:
> I would take inspiration from https://doc.rust-lang.org/std/keyword.return.html so I would like to return an expression. See [here](https://github.com/vectordotdev/vector/blob/v0.36.0/src/conditions/vrl.rs#L101-L106), we can use both the returned value and the target `.`.

I'm not sure what you mean, would you like to have only `return expr` and no `return` without arguments?

I think there will always be a difference between Rust and VRL because the latter already implicitly uses `.` when returning at the end of a program while the former implicitly returns the result of the last expression at the end of a function. So in VRL, it makes much more sense to me to have a parameter-less return, returning the same thing as if it were the end of the program.

> To clarify a bit more: would `return foo` use the value of `foo` as the event to emit? Or would it still use `.`? If the latter, what would the "returned value" be used for?

I would expect it to emit `foo`.

jszwedko replied:
> > To clarify a bit more: would `return foo` use the value of `foo` as the event to emit? Or would it still use `.`? If the latter, what would the "returned value" be used for?
> 
> I would expect it to emit `foo`.

Likewise. I'd suggest we could allow `return` both with and without an "argument". If an "argument" is provided, then it would return that, otherwise it would return `.`. My thinking is that `return X` could be useful for VRL conditions that have to evaluate to a boolean (where you might want to `return true` or `return false` early) whereas just `return` would be more common in `remap` programs where you emit an entire event.

jszwedko commented on rfcs/2023-02-08-7496-vrl-return.md:
```diff
@@ -0,0 +1,72 @@
+- New keywords that are not currently a reserved keyword can be added to the language. This would, however, constitute a breaking change.
+- This feature can also be rejected as it does not add any functionality that cannot be currently expressed.
+
+## Outstanding Questions
```
What should happen if the `return` keyword is used in a closure like:

```
tally = {}
for_each(array!(.tags)) -> |index, value| {
    if index == 1 {
    return
  }
}
```

? My instinct is to disallow the `return` keyword inside of closures.

pront replied:
A few options:
1. Disallow `return` inside closures
2. Stop the closure execution
3. Stop the VRL program and return

jszwedko replied:
I think I like option 1 the best. We could always migrate to option 2 or 3 in the future.

mladedav replied:
I would expect option 2 because that's what most languages do. While 3 could be useful, I think that would be too surprising for many users so I would not use that with the `return` keyword.

For option 2, it also does not make sense to allow argumentless returns, because there's nothing like `.`, so I would deny them in the first iteration, if that's what we end up on.

jszwedko replied:
Agreed that option 2 is probably the most familiar to users. I was thinking of starting with option 1 purely because it'd likely be easier to implement, but if we can implement option 2 that would be preference.

jszwedko reviewed (commented):
A couple of questions, but overall this proposal makes sense to me and would be a nice usability improvement.

mladedav commented:
I've added `return <expression>` to scope and explicitly put return expressions inside closures out of scope.

Feel free to comment if I've misunderstood the discussions and you'd for example rather have the behavior inside closures defined here.

jszwedko reviewed (approved):
Thanks for the updates! I think what you have here now sounds good to me. I'd like to get @pront's review too.

pront commented:
Two concerns:

1. I am not sure about implicitly returning `.`, can we always require an expression?
2. Regarding making closures a special case, let's make sure we throw a compilation error if we detect a `return` inside the body of a closure.

mladedav commented:
Ok, I've changed the proposal to not include the implicit returned value. As long as you and @jszwedko agree on this, I have no problem with that. It's also possible to change it and add the implicit remove later without a breaking change if you change your mind later, though I don't think it's going to be an issue since `return .` vs `return` is very little overhead.

I've added a line about returns in closure being a compilation error in the user experience section.

jszwedko commented:
Thanks @mladedav !

Given allowing naked `return` statements in the future (that would return `.`) wouldn't be a breaking change, I'm ok with preceeding with simply `return <expression>` for now and considering it later to keep this unblocked.

jszwedko commented:
Thanks for this proposal @mladedav ! I realized I'm not sure: is this something you might be interested in implementing?

pront commented:
> # Regression Detector Results
> Run ID: 60b99b82-4463-4a6a-a973-50e9ea15f556 Baseline: [342b48c](https://github.com/vectordotdev/vector/commit/342b48c0f7c0aa1147a3a2a1b00089a482436560) Comparison: [2f1c785](https://github.com/vectordotdev/vector/commit/2f1c7850fbc039a894f51b844e919adf2fdc925d) Total CPUs: 7
> 
> Performance changes are noted in the **perf** column of each table:
> 
> * ✅ = significantly better comparison variant performance
> * ❌ = significantly worse comparison variant performance
> * ➖ = no significant change in performance
> 
> ## No significant changes in experiment optimization goals
> **Confidence level:** 90.00% **Effect size tolerance:** |Δ mean %| ≥ 5.00%
> 
> There were no significant changes in experiment optimization goals at this confidence level and effect size tolerance.
> 
> ## Fine details of change detection per experiment
> ## Explanation

It's funny that this check ran since we only added an `.md` file.

mladedav commented:
@jszwedko Yeah, I'd like to give it a shot. I'm just not exactly sure about timeframes and such .

jszwedko commented:
> @jszwedko Yeah, I'd like to give it a shot. I'm just not exactly sure about timeframes and such .

No rush!
